### PR TITLE
Handle memory requests via directives

### LINKED ISSUE #258: Handle job memory requests via directives
### Feature description

While working with @jdice on #256 and #257, I considered whether our memory usage requests should be handled with directives. I think this would be a good use case for directives.

### Proposed solution

Add a directive `@directives(memory='4g')` or similar. It may need to parse strings like `1024m` or `4g` and do its math in terms of gigabytes, or we could always assume units of gigabytes and only accept numeric inputs.

### Additional context

Currently memory requests are handled inconsistently across clusters/schedulers - SLURM's handling is improved by #256 and #257 but TORQUE/LSF needs to be made consistent at some point.

### PR DESCRIPTION
**Pull request by kidrahahjo**
## Description
Use the memory directive for handling memory requests in all the schedulers. The default memory reserved for each bundle is `4 gigabytes`. Users can still use the `--memory` flag with CLI. In case they specify the `--memory` flag (where `memory` is only valid when the suffix is either "g", "G",  "m", or "M") this flag will override the memory directive specified by the user. In the `memory` directive, users can only enter a numeric value in gigabytes.
## Motivation and Context
Resolves #258 

## Types of Changes

- [ ] Documentation update
- [ ] Bug fix
- [x] New feature
- [ ] Breaking change<sup>1</sup>

<sup>1</sup>The change breaks (or has the potential to break) existing functionality.

## Checklist:

- [x] I am familiar with the [**Contributing Guidelines**](https://github.com/glotzerlab/signac-flow/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**Contributor Agreement**](https://github.com/glotzerlab/signac-flow/blob/master/ContributorAgreement.md).
- [x] My name is on the [list of contributors](https://github.com/glotzerlab/signac-flow/blob/master/contributors.yaml).
- [x] My code follows the [code style guideline](https://github.com/glotzerlab/signac-flow/blob/master/CONTRIBUTING.md#code-style) of this project.
- [x] The changes introduced by this pull request are covered by existing or newly introduced tests.

If necessary:
- [ ] I have updated the API documentation as part of the package doc-strings.
- [ ] I have created a separate pull request to update the [framework documentation](https://docs.signac.io/) on [signac-docs](https://github.com/glotzerlab/signac-docs) and linked it here.
- [ ] I have updated the [changelog](https://github.com/glotzerlab/signac-flow/blob/master/changelog.txt).

### CODE REVIEW

**Comment by kidrahahjo** [comment]
@bdice @vyasr I tried covering all the environments. Please let me know how to proceed on this. I think once this gets merged, we need to update signac docs.

**Comment by vyasr** [comment]
@kidrahahjo thanks for this! I think this is the right idea. As far as deprecation, I would add a deprecation in `_main_submit` that checks whether `memory` is one of the passed kwargs. I think that's the only place where it would be necessary.

File: flow/util/template_filters.py
Comment (bdice) [inline flow/util/template_filters.py:145]: This needs to be public API.
```suggestion
def calc_memory(operations, memory=None):
```

File: flow/util/template_filters.py
Comment (bdice) [inline flow/util/template_filters.py:135]: Should this default to gigabytes? If not, we need to add a message for the error.
```suggestion
            raise ValueError("Memory amounts must be specified with a suffix "g" for gigabytes or "m" for megabytes.")
```

File: flow/templates/lsf.sh
Comment (bdice) [inline flow/templates/lsf.sh:5]: Can you verify if this is the correct syntax? I found two references online for BSUB memory requests, and both of them state that this should be given as an integer value in _kilobytes_.

https://scicomp.ethz.ch/public/manual/LSF/9.1.3/lsf_admin/index.htm?chap_interactive_jobs.html~main
http://www.glue.umd.edu/lsf-docs/man/bsub.html

Comment (bdice) [reply flow/templates/lsf.sh:5]: Submissions that do not define a memory directive should not add this line, so this needs to be set in a conditional block. See the section below with `{% if walltime %}` for an example.

File: flow/templates/slurm.sh
Comment (bdice) [inline flow/templates/slurm.sh:5]: Submissions that do not define a memory directive should not add this line, so this needs to be set in a conditional block. See the section below with `{% if walltime %}` for an example.

File: flow/templates/torque.sh
Comment (bdice) [inline flow/templates/torque.sh:4]: Submissions that do not define a memory directive should not add this line, so this needs to be set in a conditional block. See the section below with `{% if walltime %}` for an example.

Comment (bdice) [reply flow/templates/torque.sh:4]: The references I found below use `mem` instead of `pmem` and expect the suffix to be `gb` or `GB`. I'm not sure what is best to use here.

https://www.osc.edu/supercomputing/batch-processing-at-osc/pbs-directives-summary
https://docs.pace.gatech.edu/faqs/pbs/

File: flow/util/template_filters.py
Comment (bdice) [inline flow/util/template_filters.py:112]: Rename this to make it clearer that it's parsing a string to a numerical value.
```suggestion
def _parse_memory_flag(memory):
```

Comment (bdice) [reply flow/util/template_filters.py:112]: If this should be removed in 0.14 (when `--memory` is removed) then please add a deprecation notice or comment that contains the string "0.14" (that's what I'll search for when checking that deprecated code was removed).

File: flow/util/template_filters.py
Comment (bdice) [inline flow/util/template_filters.py:138]:
```suggestion
            'Invalid memory flag. For gigabytes use suffix "g", '
```

**Review by bdice — CHANGES_REQUESTED** [review]
Changes requested in comments. Overall this is a good start, we just need to iron out some scheduler-specific stuff.

File: flow/util/template_filters.py
Comment (bdice) [inline flow/util/template_filters.py:131]:
```suggestion
            return float(memory[:-1]) / 1024
```

File: flow/util/template_filters.py
Comment (bdice) [inline flow/util/template_filters.py:133]:
```suggestion
            return float(memory[:-1])
```

File: flow/templates/lsf.sh
Comment (kidrahahjo) [reply flow/templates/lsf.sh:5]: I saw this documentation which says that we can use `G` or `GB` for `gigabytes` but by default, the memory is specified in kilobytes.
https://www.ibm.com/support/knowledgecenter/SSWRJV_10.1.0/lsf_command_ref/bsub.__m.1.html

File: flow/util/template_filters.py
Comment (kidrahahjo) [reply flow/util/template_filters.py:135]: The command line flag currently requires users to enter memory with units in gigabytes (`g`) or megabytes(`m`). In order to default it to `gigabytes`, we need to deprecate the current behaviour. (Right?)

File: flow/util/template_filters.py
Comment (kidrahahjo) [reply flow/util/template_filters.py:145]: Sounds good!

File: flow/templates/torque.sh
Comment (kidrahahjo) [reply flow/templates/torque.sh:4]: For the command line flag, we ask users to enter memory to request per node. 
`pmem` -> Memory to request per process
`mem` -> Memory to request overall.

I am not sure what's the context of "per node" in our case. (I think it means total memory so we should use `mem`)

File: flow/templates/slurm.sh
Comment (kidrahahjo) [reply flow/templates/slurm.sh:5]: I agree that we should add it only when users ask us to but we define a default value for the memory directive [here](https://github.com/glotzerlab/signac-flow/blob/master/flow/directives.py#L380), I thought this was done in order to actually default the memory requests to 4 gigabytes. Can you please confirm this?

Comment (bdice) [reply flow/templates/slurm.sh:5]: Good observation. signac-flow should not define a default memory amount because (a) it breaks backwards compatibility and (b) doesn't match how schedulers would normally act. We should set the directive's default to `None`. I'm not sure off the top of my head how walltimes work with directives when some operations define a walltime and others do not, but that would be a good reference. I propose that accumulation would be done like this:
- In serial, we take the max. The max of a defined amount (e.g. 4 GB) and an undefined amount (`None`) should be the defined amount (4 GB).
- In parallel, we take the sum. The sum of a defined amount (e.g. 4 GB) and an undefined amount (`None`) should be the defined amount (4 GB).

File: flow/templates/lsf.sh
Comment (bdice) [reply flow/templates/lsf.sh:5]: Excellent! That's the right reference, I just couldn't find it when I was googling. Let's use `GB` here.
```suggestion
#BSUB -M {{ operations | _calc_memory(memory) }}GB
```

File: flow/util/template_filters.py
Comment (bdice) [reply flow/util/template_filters.py:135]: Currently values without a suffix `g` or `m` are not supported. I was proposing adding a _new_ behavior (not breaking a current behavior) that would allow a numeric value with no suffix to be interpreted as gigabytes.

If you think that providing a raw number without a suffix (and interpreting it as gigabytes) _should_ be allowed, then accept the first suggestion `return float(memory)`. If you think that providing a raw number without a suffix _should not_ be allowed, then accept the second suggestion `raise ValueError(...)` because it makes the error message clear.

File: flow/templates/torque.sh
Comment (bdice) [reply flow/templates/torque.sh:4]: I agree with your assessment. Let's use `mem`, we intend to request a total amount of memory (not an amount per core requested).

**Comment by vyasr** [comment]
@bdice @kidrahahjo let me know if you would like me to review this, I am happy to provide another set of eyes on this if you think it would be helpful or if there are any questions/issues you're not able to resolve between the two of you.

**Comment by bdice** [comment]
@vyasr Thanks for the offer, I'd really appreciate your input on this PR when @kidrahahjo is ready for review!

**Comment by kidrahahjo** [comment]
@bdice @vyasr I think this is ready now, please have a look.

File: flow/directives.py
Comment (kidrahahjo) [inline flow/directives.py:272]: I am not really satisfied with the logic I used here. But since it's internal and the variables which use this are internal as well, I am not really sure if we should be too worried about this.

Comment (vyasr) [reply flow/directives.py:272]: I think this code is fine. It looks identical to the place in hoomd-blue where I assume @b-butler [originally advocated copying it from](https://github.com/glotzerlab/hoomd-blue/blob/master/hoomd/data/typeconverter.py#L153).

File: flow/directives.py
Comment (vyasr) [inline flow/directives.py:257]: Since this class now accept multiple types, `_OnlyTypes` would be a more appropriate name.

File: flow/directives.py
Comment (vyasr) [inline flow/directives.py:338]: I would add a comment that 1e-12 is an arbitrarily chosen minimum threshold for what constitutes 0.

File: flow/directives.py
Comment (vyasr) [inline flow/directives.py:384]: Is there a reason not to support units other than Gb? I see that the function for parsing of string memory flags is marked for deprecation because we'll be removing the `--memory` CLI arguments, but we could still support passing a string here. I have a slight preference for supporting this since it's a more natural mapping to the interface that users are accustomed to when manually writing cluster submission scripts, and I don't think there's much overhead to doing so (we can still default to Gb if no units are provided).

Comment (vyasr) [reply flow/directives.py:384]: This does make the code slightly more complex since the templates for different schedulers can't assume a suffix. I'm fine skipping this for now if you can't find a clean and simple solution, we can always add it later if people want that support.

File: flow/project.py
Comment (vyasr) [inline flow/project.py:4511]: Use warnings.warn with a DeprecationWarning.

File: flow/util/template_filters.py
Comment (vyasr) [inline flow/util/template_filters.py:199]: I don't think this logic is quite right. The `--memory` command-line argument's help says that this is the amount of memory to request *per node*. That is also likely why the old torque.sh template uses `pmem` instead of `mem`. Unfortunately, this complicates our logic quite a bit because we don't usually know how many nodes we will have a priori, and it's a fundamentally different calculation than the per operation calculation. Since the `--memory` flag is command line, and I suspect that very few people are using it anyway, I think we may be better off skipping a deprecation cycle for it and just removing it rather than trying to support two incompatible measures of memory.

**Review by vyasr — CHANGES_REQUESTED** [review]
You can look through my inline comments, but there is one significant change I'm suggesting that may make some of my comments inapplicable. The `--memory` command-line flag is actually incompatible with the memory directive in its current form because it claims to define "memory per node". We are not even consistent about supporting this correctly between different schedulers (lsf doesn't seem to have had any memory support, slurm treated the memory as a total rather than per-node, and torque did what we claimed it should), and this calculation can't be nicely mapped to the per-operation directive introduced by this PR. My suggestion would be to skip the deprecation cycle in this case and just remove support for the CLI memory entirely. It's a very minor break in the CLI that saves us significant problems here _and_ makes our execution model better defined going forward.

File: flow/project.py
Comment (kidrahahjo) [reply flow/project.py]: This whole block is removed hence this change is no longer required

File: flow/directives.py
Comment (kidrahahjo) [reply flow/directives.py]: We now support

1. Positive numeric value with suffix "g" or "G" indicating memory requested in gigabytes.
2. Positive numeric value with suffix "m" or "M" indicating memory requested in megabytes.
3. Positive numeric value with no suffix indicating memory requested in gigabytes.

File: flow/util/template_filters.py
Comment (vyasr) [reply flow/util/template_filters.py:138]: I thought that all instances of the period before the `(Default value = ...)` were created by Bradley during all his refactoring and performance optimization work, but perhaps some were yours as well? In any case, period after is much more commonly used in `signac` (and other projects in the `glotzerlab` organization that use Google docstrings, for instance `freud`, `rowan`, and `coxeter`). It would be nice to make this consistent everywhere... but it's probably the lowest priority task every so I'll just point it out here for future changes 😂 .

File: tests/test_util.py
Comment (vyasr) [inline tests/test_util.py:62]: You could define these as fixtures then reuse them in `test_check_memory` below.

**Review by vyasr — APPROVED** [review]
The code looks good to me now. Two main notes:
1. I'm not sure why the `template_reference_data.tar.gz` file shows up in the diff as empty. Can you double-check that the file has contents? I assume it's fine since tests are passing, but please just check again to be sure.
2. There are a couple of code paths missing coverage in `flow/directives.py`. It would be nice to add coverage for them.

The coverage issue isn't a big deal and IMO not worth blocking this PR for, @bdice feel free to comment when you review though. Before approving/merging, can you also make sure that the first issue (empty template reference data file) has been resolved one way or another. Aside from that I am approving this PR.

**Comment by bdice** [comment]
@vyasr @kidrahahjo I think the template reference data appearing to have zero size is a bug in GitHub, and not a problem with this PR. I saw that on the last couple PRs that altered template reference data, but the file appeared just fine. You can check it just to be sure but don’t worry too much about it.

File: flow/util/template_filters.py
Comment (bdice) [reply flow/util/template_filters.py:112]: Above, I suggested removing `check_memory` and simply checking the value of `calc_memory` in the template (`None` if no operations specify a memory directive). That would eliminate the need for Vyas's suggestion.

**Review by bdice — CHANGES_REQUESTED** [review]
The PR looks good overall but can be simplified a bit by removing `check_memory` and only using `calc_memory`. I may try to make a pass on this PR on Monday/Tuesday to apply the suggested changes and help finish it up.

File: tests/test_util.py
Comment (kidrahahjo) [reply tests/test_util.py]: Test check_memory is no longer needed but I'll still define them as fixtures.

File: flow/util/template_filters.py
Comment (kidrahahjo) [reply flow/util/template_filters.py]: The use of `check_memory` has now been removed.

**Pull request merged**
